### PR TITLE
Feature/add/user management

### PR DESCRIPTION
**Pull request by vjumpkung**

### CODE REVIEW

File: src/schema/users.schema.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { Types } from 'mongoose';
 
 export enum UserRole {
   Owner = 100,
+  Chef = 75,
```
Comment (ishiasaka) [inline src/schema/users.schema.ts]: Why have Chef's in UserRole, as we discussed chef == employee

File: src/users/dto/user.update.role.dto.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import { ApiProperty } from '@nestjs/swagger';
+import { Transform } from 'class-transformer';
+
+export class UpdateRoleDto {
+  @ApiProperty({
+    type: String,
+    enum: UserRoleString,
+    example: UserRoleString.Owner,
+    description:
+      'select role from enum UserRole example: Owner, Chef, Cashier, Employee, Customer',
```
Comment (ishiasaka) [inline src/users/dto/user.update.role.dto.ts]: Why have Chef's in UserRole, as we discussed chef == employee

File: src/users/users.service.ts [hunk 3]
```diff
@@ -73,4 +75,39 @@ export class UsersService {
       })
       .exec();
   }
+
+  async updateRole(id: Types.ObjectId, updateRole: UpdateRoleDto) {
+    await this.userModel
+      .findByIdAndUpdate(id, {
```
Comment (ishiasaka) [inline src/users/users.service.ts]: use `updateOne` if you don't want document

File: src/schema/users.schema.ts (hunk 1 above)
Comment (vjumpkung) [reply src/schema/users.schema.ts]: Okay but in README.md Chef is 75

File: src/users/users.service.ts [hunk 4]
```diff
@@ -73,4 +73,48 @@ export class UsersService {
...
+        },
+      )
+      .exec();
+  }
+
+  async updateUserInfo(userId: Types.ObjectId, updateInfo: UpdateInfoDto) {
+    const user = await this.userModel
+      .findOne({ _id: userId }, { password: true })
+      .exec();
+    if (argon2.verify(user.password, updateInfo.oldPassword)) {
```
Comment (ishiasaka) [inline src/users/users.service.ts]: Is this `Promise`?

**Review by ishiasaka — CHANGES_REQUESTED** [review]

File: src/users/users.service.ts (hunk 4 above)
Comment (vjumpkung) [reply src/users/users.service.ts]: no will await later.

File: src/users/users.service.ts [hunk 5]
```diff
@@ -73,4 +73,48 @@ export class UsersService {
...
+        const hashedPassword = await argon2.hash(updateInfo.newPassword);
+        await this.userModel
+          .updateOne(
+            { _id: userId },
+            {
+              password: hashedPassword,
+            },
+          )
+          .exec();
+      }
```
Comment (lazykern) [inline src/users/users.service.ts]: So there will be 2 queries if I wanna update both username and password right?
Can they both be updated with only one query?

File: src/users/users.controller.ts [hunk 6]
```diff
@@ -92,8 +116,70 @@ export class UsersController {
...
+  })
+  @ApiBearerAuth()
+  @Role(UserRole.Owner)
+  @HttpCode(HttpStatus.NO_CONTENT)
+  async updateUserRole(
+    @User() { id }: UserJwt,
+    @Param('id', new ParseMongoIdPipe()) UpdateId: Types.ObjectId,
+    @Body() updateRoleDto: UpdateRoleDto,
+  ) {
+    if (id === UpdateId.toString()) {
```
Comment (lazykern) [inline src/users/users.controller.ts]: Use `UpdateId.equals(id)` instead

File: src/users/users.controller.ts [hunk 7]
```diff
@@ -74,9 +88,19 @@ export class UsersController {
...
+  @ApiBearerAuth()
+  @Role(UserRole.Owner)
   @HttpCode(HttpStatus.NO_CONTENT)
-  async deleteUser(@Query('id', new ParseMongoIdPipe()) id: Types.ObjectId) {
-    await this.usersService.deleteUser(id);
+  async deleteUser(
+    @User() { id }: UserJwt,
+    @Query('id', new ParseMongoIdPipe()) deleteId: Types.ObjectId,
+  ) {
+    if (id === deleteId.toString()) {
```
Comment (lazykern) [inline src/users/users.controller.ts]: Use `deleteId.equals(id)` instead

File: src/users/users.controller.ts [hunk 8]
```diff
@@ -92,8 +116,70 @@ export class UsersController {
...
+  @ApiBearerAuth()
+  @Role(UserRole.Owner)
   @HttpCode(HttpStatus.NO_CONTENT)
-  async updateUser(@Query('id', new ParseMongoIdPipe()) id: Types.ObjectId) {
-    await this.usersService.resetPassword(id);
+  async updateUser(
+    @User() { id }: UserJwt,
+    @Query('id', new ParseMongoIdPipe()) resetId: Types.ObjectId,
+  ) {
+    if (id === resetId.toString()) {
```
Comment (lazykern) [inline src/users/users.controller.ts]: Use `resetId.equals(id)` instead

**Review by ishiasaka — APPROVED** [review]
LGTM

**Pull request merged**